source/common/http/utility.cc

```diff
@@ -747,5 +750,116 @@ std::string Utility::PercentEncoding::decode(absl::string_view encoded) {
   return decoded;
 }
 
+const Utility::AuthorityAttributes Utility::parseAuthority(const absl::string_view& authority,
+                                                           uint32_t default_port) {
+  Utility::AuthorityAttributes auth_attr;
+  auth_attr.port = default_port;
+  auth_attr.host = authority;
+
+  bool is_ipv4 = true;
+  bool is_ipv6 = false;
+  bool have_port_ipv4 = false;
+  bool have_port_ipv6 = false;
+
+  // This section applies first-match-wins algorithm described from
+  // RFC3986(https://tools.ietf.org/html/rfc3986#section-3.2.2). Host literal include IP-literal,
+  // IPv4 Address or reg-name. In this function, We consider that IP-literal and IPv4 address as
+  // pure IP Address and return is_ip_address as true.
+
+  // check if passed authority matches IP-literal. IP-literal is constructed from bracket symbols
+  // and IPv6 address or future version of ip address. In this case, future version of ip address
+  // should be ignored
+  if (authority.find("[") != absl::string_view::npos) {
```

Comment:
I think per RFC the "[" should be the first character in the host name to be considered IP-literal.